src/Catalyst.Core.Modules.Keystore/KeyStoreService.cs

```diff
@@ -0,0 +1,830 @@
+        /// </summary>
+        /// <param name="keyName">
+        ///     The key name.
+        /// </param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        public async Task<byte[]> CreateCertificateAsync(string keyName,
+            CancellationToken cancel = default)
+        {
+            var cert = await CreateBcCertificateAsync(keyName, cancel).ConfigureAwait(false);
+            return cert.GetEncoded();
+        }
+
+        /// <summary>
+        ///     Create a X509 certificate for the specified key.
+        /// </summary>
+        /// <param name="keyName">
+        ///     The key name.
+        /// </param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        public async Task<X509Certificate> CreateBcCertificateAsync(string keyName,
+            CancellationToken cancel = default)
+        {
+            // Get the BC key pair for the named key.
+            var ekey = await Store.TryGetAsync(keyName, cancel).ConfigureAwait(false);
+            if (ekey == null)
+            {
+                throw new KeyNotFoundException($"The key '{keyName}' does not exist.");
+            }
+
+            AsymmetricCipherKeyPair kp = null;
+            UseEncryptedKey(ekey, key => { kp = GetKeyPairFromPrivateKey(key); });
+
+            // A signer for the key.
+            var ku = new KeyUsage(KeyUsage.DigitalSignature
+              | KeyUsage.DataEncipherment
+              | KeyUsage.KeyEncipherment);
+            ISignatureFactory signatureFactory = null;
+            switch (kp.Private)
+            {
+                case ECPrivateKeyParameters _:
+                    signatureFactory = new Asn1SignatureFactory(
+                        X9ObjectIdentifiers.ECDsaWithSha256.ToString(),
+                        kp.Private);
+                    break;
+                case RsaPrivateCrtKeyParameters _:
+                    signatureFactory = new Asn1SignatureFactory(
+                        PkcsObjectIdentifiers.Sha256WithRsaEncryption.ToString(),
+                        kp.Private);
+                    break;
+                case Ed25519PrivateKeyParameters _:
+                    signatureFactory = new Asn1SignatureFactory(
+                        EdECObjectIdentifiers.id_Ed25519.Id,
+                        kp.Private);
+                    ku = new KeyUsage(KeyUsage.DigitalSignature);
+                    break;
+            }
+
+            if (signatureFactory == null)
+            {
+                throw new NotSupportedException($"The key type {kp.Private.GetType().Name} is not supported.");
+            }
+
+            // Build the certificate.
+            var dn = new X509Name($"CN={ekey.Id}, OU=keystore, O=ipfs");
+            var ski = new SubjectKeyIdentifier(Base58.Decode(ekey.Id));
+
+            // Not a certificate authority.
+            // TODO: perhaps the "self" key is a CA and all other keys issued by it.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)